examples/echo/main.go

```diff
@@ -5,64 +5,165 @@ command.
 package main
 
 import (
+	"context"
 	"fmt"
 	"os"
 	"os/signal"
+	"sync"
 	"syscall"
+	"time"
 
 	"github.com/moderepo/device-sdk-go/v2"
 )
 
-func doEcho(_ *mode.DeviceContext, cmd *mode.DeviceCommand) {
-	// Command parameters are JSON encoded and can be retrieved as follows.
-	var params struct {
-		Msg string `json:"msg"`
-	}
+var (
+	// My cloudmqtt.com server)
+	modeMqttHost = "staging-api.corp.tinkermode.com"
+	modeMqttPort = 8883
+	modeUseTLS   = true
+)
+
+func pingLoop(client *mode_client.MqttClient, timeout time.Duration,
+	pingRecv <-chan bool, gotDisconnected chan<- bool, doDisconnect <-chan bool,
+	wg sync.WaitGroup) {
+
+	wg.Add(1)
+	defer func() {
+		fmt.Println("Exiting ping loop")
+		wg.Done()
+	}()
+	ticker := time.NewTicker(3 * time.Second)
+
+	for {
+		if err := client.Ping(); err != nil {
+			gotDisconnected <- true
+			return
+		}
+
+		// Wait for the ack, or timeout
+		ctx, cancel := context.WithTimeout(context.Background(), timeout)
+		defer cancel()
 
-	if err := cmd.BindParameters(&params); err != nil {
-		fmt.Printf("Failed to bind command parameters: %s\n", err.Error())
-		return
+		// Block on the return channel or timeout
+		select {
+		case <-doDisconnect:
+			fmt.Println("runner told me to exit")
+			return
+		case ret := <-pingRecv:
+			// There's not really a way to return false, but, since it's bool,
+			// we'll check
+			fmt.Println("Got ping response")
+			if !ret {
+				fmt.Println("sender closed ping channel")
+				gotDisconnected <- true
+			}
+		case <-ctx.Done():
+			fmt.Println("ping timeout")
+			gotDisconnected <- true
+		}
+
+		fmt.Println("Ping sent and ack'ed")
+		// Wait for the timer to before we loop again
+		<-ticker.C
 	}
+}
 
-	eventData := map[string]interface{}{"msg": params.Msg}
+func waitForAck(delegate *mode_client.ModeMqttDelegate) uint16 {
```

Comment:
`ModeMqttDelegate` method receivers are using a value. But this method is using a pointer. 
Could you tell me what is reason. Just coping cost?

Reply:
Yes. In my habit, I would have used "const &" from C++. In this case, I should just pass the channel.